src/enumerable/enumerable_gather.jl

```diff
@@ -0,0 +1,75 @@
+struct EnumerableGather{T,S,F,I} <: Enumerable
+    source::S
+    fields::F
+    indexFields::I
+    key::Symbol
+    value::Symbol
+end
+
+struct Not{T}
+    val::T
+end
+
+function gather(source::Enumerable, args...; key::Symbol = :key, value::Symbol = :value)
+    T = eltype(source)
+    fields = fieldnames(T)
+    F = typeof(fields)
+    if args != (false,)
+        prev = ()
+        firstArg = true
+        for arg in args
+            if typeof(arg) == Symbol
+                prev = (prev..., arg)
+            else typeof(arg) == Not{Symbol}
+                if firstArg
+                    prev = (a for a in fields if a != arg.val)
+                else
+                    prev = (a for a in prev if a != arg.val)
+                end
+            end
+            firstArg = false
+        end
+    else
+        prev = fields
+    end
+
+    I = typeof(prev)
+    return EnumerableGather{T, typeof(source), F, I}(source, fields, prev, key, value)
+end
+
+function Base.iterate(iter::EnumerableGather{T, S, F, I}) where {T, S, F, I}
+    source = iter.source
+    fields = fieldnames(T)
+    elements = Array{Any}(undef, 0)
+    
+    savedFields = (n for n in iter.fields if !(n in iter.indexFields))
+    for i in iter.source
+        for j in fields
+            if j in iter.indexFields
+                push!(elements, NamedTuple{(iter.key, iter.value, savedFields...)}((j, i[j], Base.map(n->i[n], savedFields)...)))
+            end
+        end
+    end
```

Comment:
I think this is going to be quite inefficient because it is almost certainly not type stable. To get around that, we probably have to use generated functions...

There is a bigger question, though: If you look at the classification [here](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/classification-of-standard-query-operators-by-manner-of-execution), then this implementation now is deferred and non-streaming. I *think* one could actually implement a deferred and streaming version. If we can, we should go with that.

The broad idea there might be that the state that we return from the ``iterate`` functions only contains say two indices: one index that references the source row, and one the source column that should be returned next as a full row. And then in each ``iterate`` call, just one element for the output stream is created.

Let me know if this is confusing, we can discuss it more on Slack or when we next meet.